# Exploring Synthetic Genomes for Genome-wide Association Studies

Codebase accompanying the Bioinformatics submission “Exploring Synthetic Genomes for Genome-wide Association Studies”

## Project Structure

- `1kgp/`: train-test-ref datasets from 1kGP 
- `models/`: CTGAN and DNADiffusion training pipelines
- `evaluation/`: Utility and privacy evaluation modules

## Datasets

This project supports two main public datasets:

- **1000 Genomes Project**  
- **OpenSNP**  
  *Note: The OpenSNP dataset was publicly available at the time of this study. The platform was closed in April 2025. All data used in this work were collected prior to its discontinuation.*




A preliminary version of this paper presenting two alternative approaches was published in [25]. This article significantly
extends that work developing an additional hybrid approach
which outperforms both approaches in [25]. Additionally, the
article includes a thorough theoretical analysis as well as an
enhanced comparative analysis with a more robust empirical
evaluation with additional baselines.


Building on our preliminary study \citep{}, this work substantially broadens the scope of investigation and provides the first comprehensive evaluation of the synthetic SNP (Single Nucleotide Polymorphism) sequences generated by state-of-the-art generative models, focusing on their ability to replicate real genomic data while maintaining privacy in the context of GWAS.

Build on our prelimariy work \citep{}, this work significantly the scope of investigation and provides the first comprehensive evaluation of the synthetic SNP (Single Nucleotide Polymorphism) sequences generated by state-of-the-art generative models, focusing on their ability to replicate real genomic data while maintaining privacy in the context of GWAS. Previous work \citep{oprisanu2021utility} has primarily focused on traditional statistical methods \citep{samani2015quantifying} and deep generative models (DGMs), such as WGAN \citep{killoran2017generating}, which struggle with the discrete nature of genomic data and rely on low-order statistics for utility measurements. In contrast, we evaluate two generative models that can handle discrete, high-dimensional data, the key characteristics of genomic data. We employ a comprehensive suit of metrics, including low-order statistics, t-SNE visualizations \citep{tsne}, and performance on two GWAS downstream tasks, four state-of-the-art membership inference attacks (MIAs), and two memorization measurements, to assess the fidelity, utility, and privacy of synthetic genomic data.